libportal/portal.c

```diff
@@ -273,3 +280,200 @@ xdp_portal_new (void)
 {
   return g_object_new (XDP_TYPE_PORTAL, NULL);
 }
+
+/**
+ * xdp_portal_running_under_flatpak:
+ * @error: return location for a #GError pointer
+ *
+ * If you are trying to determine whether to use portals, consider using
+ * xdp_portal_running_under_sandbox() instead.
+ *
+ * Returns: %TRUE if the current process is running under a Flatpak sandbox, or
+ *   %FALSE if either an error was encountered or the process is running
+ *   unsandboxed
+ */
+gboolean
+xdp_portal_running_under_flatpak (GError **error)
+{
+  static _Thread_local gboolean decided = FALSE;
+  static _Thread_local gboolean under_flatpak = FALSE;
```

Comment:
You could use `GPrivate` for thread-specific data.

Reply:
Is there an advantage to doing so? I copied `_Thread_local` from epiphany's codebase, and the `GPrivate` documentation says it is "a very limited resource", so I don't know what the reason to switch would be.

Reply:
I think `GPrivate` predates `_Thread_local`. They're surely the same thing.

Reply:
You can include `<threads.h>` and have a nicer looking `thread_local`.

Also you should add checks to the build system for this to have a nice error in case somebody uses a non-c11 compiler.

Reply:
(I kinda think depending on C11 for this is a bit silly, but its also not new...)

Reply:
I'm actually not sure this should be thread-local also. Like these do blocking IO in the first place so its not like waiting on a mutex is a problem, and we want every thread to have the exact same result every time.

Reply:
The only point of thread local is to avoid mutex locking. So if you're doing I/O anyway, there's indeed no point: just use a GOnce.

Reply:
I don't understand what makes `GOnce` better than `_Thread_local` in this case?

Reply:
`_Thread_local` is a precious resource that has to be used sparingly.

You originally copied this from Epiphany, where it doesn't matter so much, but now this is a library that will be used in loads of applications.

Reply:
And `GOnce` will let the check run again if an unsandboxed process forks off a process that is sandboxed? Maybe that's a dumb question because I don't have a good grasp on how bubblewrapped processes are created for rendering thumbnails or whatever else they're used for.

Reply:
GOnce is simply a mutex protected global variable. It is only set once in a process. Subprocesses won't be affected.

Reply:
Ok, I updated it to use GOnce. It became a bit more complex but maybe it's fine.

Reply:
As I mentioned earlier I think `g_once_init_enter()` is the cleaner method.

Sorry for the back and forth.

Reply:
I don't think that would work for `xdp_portal_running_under_snap` since we call the `get_under_snap` function in the `SNAP_STATE_ERROR` case even if it's been called before?